### PR TITLE
[C5][environment] Environment support for admin portal

### PR DESCRIPTION
**Pull request by RenukaFernando**
### Proposed changes in this pull request
- Environment feature for Admin Portal
- Login page for environment support
- Change prop `onRequestClose` to `onClose` of components in Material UI with version change.
- New UI component for redirecting
- Change loading component.
  - Replace Antd `Spin` with Material UI `CircularProgress`
- Fix https://github.com/wso2/product-apim/issues/2199

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/package.json [hunk 1]
```diff
@@ -15,17 +15,20 @@
   "author": "WSO2 Org",
   "license": "Apache License 2.0",
   "dependencies": {
-    "antd": "^2.11.2",
+    "antd": "^2.13.2",
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/package.json]: Do we need to update ant version?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/source/src/app/components/Shared/Redirecting.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            <Grid container justify={"center"} alignItems={"center"} spacing={0} style={{height: "100vh"}}>
+                <Grid item lg={6} md={8} xs={10}>
+                    <Grid container alignItems={"center"}>
+                        <Grid item sm={2} xs={12}>
+                            <CircularProgress style={{float: "right"}}/>
+                        </Grid>
+                        <Grid item sm={10} xs={12}>
+                            <div className="redirect-main-content">
+                                <Paper elevation={5} square={true} className="redirect-paper"
+                                       style={{fontSize: "medium", padding: "15px"}}>
```
Comment (chamilaadhi) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/source/src/app/components/Shared/Redirecting.js]: Shall we move these styles to css file.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Shared/Redirecting.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                <Grid item lg={6} md={8} xs={10}>
+                    <Grid container alignItems={"center"}>
+                        <Grid item sm={2} xs={12}>
+                            <CircularProgress style={{float: "right"}}/>
+                        </Grid>
+                        <Grid item sm={10} xs={12}>
+                            <div className="redirect-main-content">
+                                <Paper elevation={5} square={true} className="redirect-paper"
+                                       style={{fontSize: "medium", padding: "15px"}}>
+                                    {message}
```
Comment (chamilaadhi) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Shared/Redirecting.js]: shall we move these styles to the css file.

Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Shared/Redirecting.js]: Done with https://github.com/wso2/carbon-apimgt/pull/4890/commits/9d078d50d5a10b44f5065d6c8b66ea080d11c65d

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/source/src/app/components/Shared/Redirecting.js (hunk 2 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/source/src/app/components/Shared/Redirecting.js]: Done with https://github.com/wso2/carbon-apimgt/pull/4890/commits/9d078d50d5a10b44f5065d6c8b66ea080d11c65d

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/package.json (hunk 1 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/package.json]: Revert with https://github.com/wso2/carbon-apimgt/pull/4890/commits/32e120551eb5cd6b154f8eb65ad7c3683c239716

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+class EnvironmentMenu extends React.Component {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            openEnvironmentMenu: false
+        };
+    }
+
+    handleClickEnvironmentMenu = event => {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: Shall we avoid using ES7 class properties, As stated in react [documentation](https://reactjs.org/docs/handling-events.html) this is an experimental feature. If we use this syntax we have to use this babel [plugin](https://babeljs.io/docs/plugins/transform-class-properties/) which we don't use.

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        this.setState({openEnvironmentMenu: false});
+        //TODO: [rnk] Temporary Fix: Reload the web page when environment changed
+        document.location.reload();
+    };
+
+    render() {
+        //Props list
+        const environments = this.props.environments;
+        const environmentLabel = this.props.environmentLabel;
+        const handleEnvironmentChange = (event) => {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: AFAIK there is a small performance impact on creating arrow function when always rendering the element, shall we move this function to class level

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                <Menu
+                    id="simple-menu"
+                    anchorEl={this.state.anchorElEnvironmentMenu}
+                    open={this.state.openEnvironmentMenu}
+                    onRequestClose={this.handleRequestCloseEnvironmentMenu}
+                    style={{alignItems: "center", justifyContent: "center"}}
+                >
+                    {environments.map((environment, index) =>
+                        <MenuItem onClick={handleEnvironmentChange}
+                                  id={index}>{environment.label}</MenuItem>
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: styling issue

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/AuthManager.js [hunk 7]
```diff
@@ -178,18 +191,19 @@ class AuthManager {
...
             'Content-Type': 'application/json',
             'Authorization': authHeader
         };
         const promisedLogout = axios.post(url, null, {headers: headers});
         return promisedLogout.then(response => {
-            Utils.delete_cookie("WSO2_AM_TOKEN_1");
-            localStorage.removeItem("wso2_user");
+            Utils.delete_cookie(User.CONST.WSO2_AM_TOKEN_1, Utils.CONST.CONTEXT_PATH);
+            localStorage.removeItem(User.CONST.LOCALSTORAGE_USER);
+            new APIClientFactory().getAPIClient(Utils.getEnvironment().label)._instance = null; // Single client should be re initialize after log out
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/AuthManager.js]: I think this is not correct, if `._instance` intended to be a private variable in APIClient class, you should not set it in this manner, use a class method to flush the current APIClient instance for given environment.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js [hunk 8]
```diff
@@ -47,11 +47,30 @@ const Logout = () => import(/* webpackChunkName: "logout" */ './app/components/L
...
-        this.state = {showLeftMenu: false};
+        this.state = {
+            showLeftMenu: false,
+            authConfigs: null
+        };
         message.config({top: '48px'}); // .custom-header height + some offset
         /* TODO: need to fix the header to avoid conflicting with messages ~tmkb*/
     }
 
+    handleResponse = (response) => {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js]: same as mentioned above, this is an ES7 syntax which is currently in proposal stage 3

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/Utils.js [hunk 9]
```diff
@@ -141,16 +142,20 @@ class Utils {
         AuthManager.setUser(user);
     }
 
-    static getAppLoginURL() {
-        return Utils.CONST.PROTOCOL + Utils.getEnvironment().host + Utils.CONST.LOGIN + Utils.CONST.CONTEXT_PATH;
+    static getPromised_ssoData(environment){
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/Utils.js]: styling issue 
```javascript
){
```

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Login/Login.js [hunk 10]
```diff
@@ -41,129 +47,188 @@ class Login extends Component {
...
+        let loginPromise = this.authManager.authenticateUser(username, password, environment);
         loginPromise.then((response) => {
             this.setState({isLogin: AuthManager.getUser(), loading: false});
         }).catch((error) => {
-                this.setState({ messageOpen: true });
-                this.setState({message: error});
+                this.setState({messageOpen: true});
+                if(error.response.data.code === 900964){ //Error while generating a new access token.
+                    this.setState({message: "Invalid Username or Password."});
+                }else{
```
[the remark this answers is not in the document]
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Login/Login.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/abbccb51b99a9e06ff1cded6c2cbf63c2ead62dd

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js (hunk 4 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: Done with https://github.com/wso2/carbon-apimgt/pull/4890/commits/57e9929cdc68c5b4447681c94a793b86b0390380

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js (hunk 6 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/abbccb51b99a9e06ff1cded6c2cbf63c2ead62dd

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/Utils.js (hunk 9 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/Utils.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/abbccb51b99a9e06ff1cded6c2cbf63c2ead62dd

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js (hunk 8 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/App.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/57e9929cdc68c5b4447681c94a793b86b0390380

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js (hunk 5 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Base/Header/EnvironmentMenu.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/57e9929cdc68c5b4447681c94a793b86b0390380

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/AuthManager.js (hunk 7 above)
Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/data/AuthManager.js]: Fixed with https://github.com/wso2/carbon-apimgt/pull/4890/commits/889b3933af3afb8baa807e737d042ba2c27b27a3

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Login/Login.js [hunk 11]
```diff
@@ -91,13 +91,13 @@ class Login extends Component {
...
-                if(error.response.data.code === 900964){ //Error while generating a new access token.
+                if (error.response.data.code === 900964) { //Error while generating a new access token.
                     this.setState({message: "Invalid Username or Password."});
-                }else{
+                } else {
                     this.setState({message: error.message});
                 }
                 console.log(error);
-                this.setState({loading: false});
+                this.setState({loading: false});q
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Login/Login.js]: @RenukaFernando , Please remove this un-intended character `q` if this is present in current code

Comment (RenukaFernando) [reply features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/Login/Login.js]: Noted. Thanks.

File: features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/public/css/animations.css [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+/* New custom css */
+.social_links img{
+    width: 32px;
+    height: 32px;
+    margin-right: 5px;
+}
+.tab-container{
+    padding: 40px;
+}
```
Comment (ruwiniwj) [inline features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/public/css/animations.css:136]: add new line eof

**Pull request merged**